### ISSUE TITLE
Sysrepo apply_change timeout after 10 seconds with 50 seconds timeout provided

### ISSUE THREAD

**Issue by Krisscut** [sysrepo/sysrepo#3471]
**Versions:**

sysrepo-2.2.150
libnetconf2-3.0.11.devel
libyang-2.1.148
sysrepo-python-1.5.0
netopeer2-2.2.13

**Description of the issue**

During startup of our apps, we create some users and at some point, we fill the ietf-netconf-server 's call home and listen entries.
To do so, our app (CallHomeManager) uses the sysrepo-python method **apply_changes**, which calls the underlying sysrepo **sr_apply_changes** method.

This change triggers an **ietf-netconf-notifications" "notif"** with ID 9 (tracked in cyan in the following screenshots), which is then processed by the various subscribers.
But the **orusw**, a C application using **libsysrepo** directly, takes a bit of a time to process the notification, because it needs to create linux users, and in our case it took it 14 seconds to create it. 

The issue is that it seems the **apply_change** operation times out after only 10 seconds, as visible in those screen caps:

![image](https://github.com/user-attachments/assets/41232f64-2f0c-4ba8-a164-51684ef7fd4f)
![image](https://github.com/user-attachments/assets/22cd642f-bc38-40a2-8f3c-319edb02087c)

But inside our code, we provided a timeout of 50 seconds.

![image](https://github.com/user-attachments/assets/45111893-527a-47e7-afe4-e039361572b2)

Since the **sysrepo-python** is a wrapper around the sysrepo lib, the timeout must have originated from the underlying sysrepo code.
Do you know if another timeout could be raised distinct from the main timeout that got provided in the **apply_change** function ?

I was thinking of maybe patching the lib to add some debug traces to understand where the timeout is raised, but do you have other advise to debug this ?

[extract_timeout_fault.txt](https://github.com/user-attachments/files/18041473/extract_timeout_fault.txt)

**Comment by irfanHaslanded** [comment]
The timeout you see is 
```c
/** timeout for locking subscription SHM; maximum time an event handling should take (ms) */
#define SR_SUBSHM_LOCK_TIMEOUT 10000
```

The error seems to be coming from [sr_shmsub_notify_new_wrlock]( https://github.com/sysrepo/sysrepo/blob/v2.2.150/src/shm_sub.c#L390), which is likely called by `sr_modinfo_generate_config_change_notif` from `sr_changes_notify_store`.

Is it possible that multiple changes are being triggered from different processes/threads? 

```
2024-11-27 11:37:24.830 INFO/UE_APP-OAM/CallHomeManager: util.py:56 - log_callback() - EV ORIGIN: "ietf-netconf-notifications" "notif" ID 9 priority 0 for 7 subscribers published.

```
vs 

```
2024-11-27 11:37:34.842 ERROR/UE_APP-OAM/CallHomeManager: util.py:56 - log_callback() - Waiting for subscription of "ietf-netconf-notifications" failed, previous event "notif" ID 9 was not processed.
2024-11-27 11:37:34.843 ERROR/UE_APP-OAM/CallHomeManager: util.py:56 - log_callback() - Failed to generate netconf-config-change notification, but changes were applied.
```

Maybe as a first step, you can try the latest version of sysrepo, since your version is over 10 months old, and this code has multiple fixes since.

**Comment by Krisscut** [comment]
Thanks for your quick reply !

We have a task to move to the new netopeer2 version but it is pending some other work which depends on the libyang lib as well.

As for the other lead, the code in "CallHomeManager" applychanges that were prepared through the setItem call, which under the hood calls **sr_set_item_str** :  https://github.com/sysrepo/sysrepo-python/blob/956e2142a34ee251df313e81e4784f4a0553f197/sysrepo/session.py#L1211

is it possible the multiple notifications could be triggered by the same **apply_change** call ? (if we did multiple set_items earlier) 
Or in this case it would generate a single notification with for instance the 3 items together inside ?

Because for this error message, from the logs it seems to me it is the CallHomeManager itself which triggered the event ID 9, and I don't think we have other operations going at the same time in a separate thread in this component.

> 2024-11-27 11:37:34.842 ERROR/UE_APP-OAM/CallHomeManager: util.py:56 - log_callback() - Waiting for subscription of "ietf-netconf-notifications" failed, previous event "notif" ID 9 was not processed.
> 2024-11-27 11:37:34.843 ERROR/UE_APP-OAM/CallHomeManager: util.py:56 - log_callback() - Failed to generate netconf-config-change notification, but changes were applied.

**Comment by irfanHaslanded** [comment]
`sr_set_item_str` only stages changes in the session.

Sending of notification is only started as part of `sr_apply_changes()->sr_changes_notify_store()` or some other API but probably not relevant in your case AFAICT.

It maybe helpful, if you can dump the thread ID with each sysrepo log message, or a message each time `sr_apply_changes` is invoked..

**Comment by Krisscut** [comment]
> It maybe helpful, if you can dump the thread ID with each sysrepo log message, or a message each time sr_apply_changes is invoked..

Good point I will try to reproduce with those to try to understand if the second call is somehow from another thread.

> sr_set_item_str only stages changes in the session.
> 
> Sending of notification is only started as part of sr_apply_changes()->sr_changes_notify_store() or some other API but probably not relevant in your case AFAICT.
> 

Yes, but my point is that in my case, multiple sr_set_item_str were done prior to the sr_apply_changes.
Does it bundles them in one single notification message once the apply_change is done or does it try to send multiples ones ? (and in parallel or one after the other ?) 
It was wondering if maybe second change notification is stuck trying to claim a lock taken by the previous one still being processed.

**Comment by Krisscut** [comment]
Hi !

I added the thread-name/thread id to our python services, but issue is quite hard to reproduce.
I did manage to encounter a similar issue, attaching an extract of the traces again in a separate file:

[log_extract_sysrepo_issue.txt](https://github.com/user-attachments/files/18290051/log_extract_sysrepo_issue.txt)

In the traces, at the end the error is as follow when trying to run a sysrepocfg command::

> Error while running cmd sysrepocfg --verbosity 3 --edit=/tmp/xmlListenTree.xml --datastore=running --format=xml. cmd output: b'[INF] Connection 62 created.\n[INF] Session 68 (user "root", CID 62) created.\n[INF] EV ORIGIN: "ietf-netconf-server" "done" ID 1 priority 0 for 1 subscribers published.\n[INF] EV ORIGIN: "ietf-netconf-server" "done" ID 1 priority 0 succeeded.\n[ERR] Waiting for subscription of "ietf-netconf-notifications" failed, previous event "notif" ID 2 was not processed.\n[ERR] Failed to generate netconf-config-change notification, but changes were applied.\nsysrepocfg error: Failed to merge edit data (Timeout expired)\n'

But tracking this notification in the log, it seems it already went wrong processing the ietf-netconf-notifications with ID1:

> 2024-12-24 05:36:10.654 ERROR/UE_APP-OAM/UserManagement: [Dummy-4/140664113186368] util.py:56 - log_callback() - Internal error (/workspace/Services/3rdParty/sysrepo-2.2.150/src/shm_sub.c:4673).
--> https://github.com/sysrepo/sysrepo/blob/d57c49551fbc177bf07451f50461bf4319764692/src/shm_sub.c#L4670-L4675

It seems that issue occurs whenever an **ietf-netconf-notifications** is received while the **UserManagement service** is starting up:
![image](https://github.com/user-attachments/assets/03183747-8904-402f-9a1a-a027d8d3c232)

The code subscribe for the notifications, then later on tries to populate some sysrepo data.
As visible in the traces, "UserManagement" has 2 threads "dummy-3" and "dummy-4", I guess one for each subscriptions.

As for notif with ID2, it seems to be processed just fine, but is never completely processed because last service handling it wrote that there is "1 remaining subscriber". But I wonder if this could be linked to sysrepo being in a bad state following failure with ID1 notif.

**Comment by michalvasko** [comment]
The specific internal error you encountered has been fixed so I really think you should focus on finishing the update and only then try to reproduce and learn more about the problem, if it still occurs.
